Template/Variable/CustomRequestProcessingVariable.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\TagManager\Template\Variable;
+
+use Piwik\Piwik;
+use Piwik\Settings\FieldConfig;
+use Piwik\Validators\NotEmpty;
+
+class CustomRequestProcessingVariable extends BaseVariable
+{
+    const ID = 'CustomRequestProcessing';
+
+    public function getId()
+    {
+        return self::ID;
+    }
+
+    public function getCategory()
+    {
+        return self::CATEGORY_OTHERS;
+    }
+
+    public function isCustomTemplate()
+    {
+        return true;
+    }
+
+    public function getParameters()
+    {
+        return array(
+            $this->makeSetting('jsFunction', "function(request){\n\treturn request;\n}", FieldConfig::TYPE_STRING, function (FieldConfig $field) {
+                $field->title = Piwik::translate('TagManager_CustomRequestProcessingVariableJsFunctionTitle');
+                $field->description = Piwik::translate('TagManager_CustomRequestProcessingVariableJsFunctionDescription');
+                $field->uiControl = FieldConfig::UI_CONTROL_TEXTAREA;
+                $field->validators[] = new NotEmpty();
+                $field->validate = function ($value) {
+                    $value = trim($value);
+                    if (strpos($value, 'function(request){') !== 0) {
```

Comment:
should we use `stripos` here to make the comparison case insensitive?

Reply:
We could, but I wanted it consistent with the existing code here: https://github.com/matomo-org/tag-manager/blob/e3c5a5ecd5b1e968f2bba98eaf1e19909ac78ebf/Template/Variable/CustomJsFunctionVariable.php#L43

so i also used `strpos`. I have no problem changing it to `stripos`, but then I'd like to change it in `CustomJsFunctionVariable.php`, too. Just let me know what you prefer 😄

Reply:
@EPNW-Eric Yes good to add it there too :+1:

Reply:
@EPNW-Eric okay but the issue here is someone can write a function like below and it can be a hard rule to write it in specific way, you can maybe, trim the string first and check the first occurrence of `function` and check `(){}` are present or not
```
function(req)
{
}
```

Reply:
The code in `CustomJsFunctionVariable.php` (see above) checks only the presence of the `function` keyword and no `(`,`)` nor `{`.

What would the rule be to implement here? What about this:
1. the keyword `function`
2. an arbitrary amount of whitespaces, carriage returns, linebreaks and tabs(` `,`\r`,`\n`,`\t`)
3. a `(`
4. an arbitrary amount of whitespaces, carriage returns, linebreaks and tabs
5. the word `request`
6. an arbitrary amount of whitespaces, carriage returns, linebreaks and tabs
7. a `)`

If you have any other suggestions/comments please let me know so I can implement it 😃.

Reply:
@EPNW-Eric We can add something like this https://3v4l.org/HXnVM
This check should be sufficient to alert the user if they are not using the correct syntax

Reply:
The  validation at https://3v4l.org/HXnVM does not check if the function is callable with an argument. But that's the difference to a `CustomJsFunctionVariable`: A `CustomRequestProcessingVariable` ***must*** have one argument, otherwise, in the browser there will be a JavaScript exception. If it wasn't for that argument we could just use the validation logic from `CustomJsFunctionVariable`.

I think validating that an argument is there is actually quite hard, so thats why in my initial approach I forced the name of the argument to `request` and furthermore also forced the whole method signature to be `function(request){`.

With respect to the coding style considarations you made and what is already there in `CustomJsFunctionVariable` I think a good compromise would be:
```php
if (strpos($value, 'function(request)') !== 0) {
```
It guarantees that `$value` takes an argument, what is the important part here, even if we have force its name to `request`, and also allows what you mentioned earlier:
```js
function(request)
{
}
```
and additionally works on onliners:
```js
function(request){}
```

Reply:
@EPNW-Eric I think we are thinking too much on this, I had another look and what you have should be okay, as we are mentioning the same in description
![Screenshot from 2023-09-13 18-22-56](https://github.com/matomo-org/tag-manager/assets/14902277/4e764a0f-e494-4686-8dde-236132104a47)